pkg/server/container_reset_windows.go

```diff
@@ -0,0 +1,177 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"encoding/json"
+	"net/url"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/oci"
+	cio "github.com/containerd/cri/pkg/server/io"
+	"github.com/containerd/cri/pkg/store"
+	containerstore "github.com/containerd/cri/pkg/store/container"
+	sandboxstore "github.com/containerd/cri/pkg/store/sandbox"
+	"github.com/containerd/typeurl"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+func (c *criService) resetContainer(ctx context.Context, container containerstore.Container) (retErr error) {
+	var (
+		id     = container.ID
+		meta   = container.Metadata
+		config = meta.Config
+		entity = log.G(ctx).WithField("containerID", id)
+	)
+
+	// Return error if container is neither created nor exited
+	switch state := container.Status.Get().State(); state {
+	case runtime.ContainerState_CONTAINER_CREATED:
+		entity.Debugf("container is already in created state")
+		return nil
+	case runtime.ContainerState_CONTAINER_EXITED:
+	default:
+		return errors.Errorf("container is in invalid state %q", criContainerStateToString(state))
+	}
+
+	// Get sandbox config from sandbox store.
+	sandbox, err := c.sandboxStore.Get(meta.SandboxID)
+	if err != nil {
+		return errors.Wrapf(err, "sandbox %q not found", meta.SandboxID)
+	}
+	if sandbox.Status.Get().State != sandboxstore.StateReady {
+		return errors.Errorf("sandbox container %q is not running", meta.SandboxID)
+	}
+
+	// Set resetting state to prevent other start/remove/reset operations against this container
+	// while it's being reset.
+	if err := setContainerResetting(container); err != nil {
+		return errors.Wrapf(err, "failed to set resetting state for container %q", id)
+	}
+	defer func() {
+		if retErr != nil {
+			// Set container to exited if fail to start.
+			if err := container.Status.UpdateSync(func(status containerstore.Status) (containerstore.Status, error) {
+				status.Pid = 0
+				status.FinishedAt = time.Now().UnixNano()
+				status.ExitCode = errorStartExitCode
+				status.Reason = errorResetReason
+				status.Message = retErr.Error()
+				return status, nil
+			}); err != nil {
+				entity.WithError(err).Errorf("failed to set reset failure state for container %q", id)
+			}
+		}
+		if err := resetContainerResetting(container); err != nil {
+			entity.WithError(err).Errorf("failed to reset resetting state for container %q", id)
+		}
+	}()
+
+	// Recreate container volatile directory (if it was removed during system restart)
+	// Root directory should not have been removed
+	// No need to defer removal
```

Comment:
why?

Reply:
For the root container, they are supposed to be non-volatile and persist across restarts.
For deferring removal, `CreateContainer` only removes it if the container fails to be created, and `StartContainer` assumes it exists if the container exists. So even if the reset fails, if the container exists, then the volatile folder should also exist.
I can't remember where I read it, but I think I remember seeing that either the volatile folder (`C:\ContainerPlatData\state`) isnt guaranteed to persist across restarts, so I am warding against that case.

Reply:
could you add the additional information about not needing to defer to your comment?